model/request/ProctorioRequestBuilder.php

```diff
@@ -0,0 +1,157 @@
+     * @throws common_exception_NotFound
+     */
+    public function build(DeliveryExecutionInterface $deliveryExecution, string $launchUrl, array $options): array
+    {
+        $this->options = $options;
+
+        return
+            [
+                //delivery execution level
+                ProctorioConfig::LAUNCH_URL => $launchUrl,
+                ProctorioConfig::USER_ID => $deliveryExecution->getUserIdentifier(),
+
+                //platform level
+                ProctorioConfig::OAUTH_CONSUMER_KEY => $this->getOauthCredentials(),
+
+                ProctorioConfig::EXAM_START => $launchUrl,
+                ProctorioConfig::EXAM_TAKE => $this->getExamUrl(),
+                ProctorioConfig::EXAM_END => $this->getExamUrl(),
+                ProctorioConfig::EXAM_SETTINGS => $this->getExamSettings(),
+
+                //delivery execution level
+                ProctorioConfig::FULL_NAME => $this->getUserFullName($deliveryExecution),
+                //Delivery level
+                ProctorioConfig::EXAM_TAG => $deliveryExecution->getDelivery()->getLabel(),
+
+                ProctorioConfig::OAUTH_TIMESTAMP => $this->getTime(),
+                ProctorioConfig::OAUTH_NONCE => $this->getNonce(),
+            ];
+    }
+
+
+    /**
+     * @param $name
+     * @return mixed|null
+     */
+    private function getOption($name)
+    {
+        return $this->options[$name] ?? null;
+    }
+
+
+    /**
+     * @return string
+     */
+    protected function getExamUrl(): string
+    {
+        return tao_helpers_Uri::url(
+            'runDeliveryExecution',
+            'DeliveryRunner',
+            null,
+            []
+        );
+    }
+
+
+    /**
+     * @param DeliveryExecutionInterface $deliveryExecution
+     * @return string
+     * @throws common_exception_Error
+     * @throws common_exception_NotFound
+     */
+    protected function getUserFullName(DeliveryExecutionInterface $deliveryExecution): string
+    {
+        /** @var User $user */
+        $user = new core_kernel_classes_Resource($deliveryExecution->getUserIdentifier());
+        $fullName = UserHelper::getUserFirstName($user) ?? '';
+        $fullName .= ' ' . UserHelper::getUserLastName($user) ?? '';
+        return $fullName;
+    }
+
+    /**
+     * @return string
+     */
+    protected function getOauthCredentials(): ?string
+    {
+        return $this->getOption(ProctorioApiService::OPTION_OAUTH_KEY);
+    }
+
+    /**
+     * @return string
+     */
+    protected function getExamSettings(): ?string
+    {
+        return $this->getOption(ProctorioApiService::OPTION_EXAM_SETTINGS);
+    }
+
+    /**
+     * @return string
+     */
+    protected function getNonce(): string
+    {
+        try {
+            $nonce = $this->getUniquePrimaryKey();
+        } catch (Throwable $exception) {
+            $this->$this->logError('UUID assignation for proctorio nonce has failed');
+            $nonce = (string)$this->getTime();
```

Comment:
May be better to go with something more unique, at least this?
```suggestion
            $nonce = \oat\tao\model\security\TokenGenerator::generate();
```

Reply:
i have 2 issues with this change
 - fist i am using it to provide the current time to the timestamp Oauth field
 - second the `generate` might throw \common_Exception